### PR TITLE
feat: performance optimization for build_header()

### LINKED ISSUE #15: Something about performance improvement
Hi there,

When dealing with HTTP2 traffic, I had to do string concatenation, so I tested the following three implementations, but found that the performance of t1k was worse than the other two. And maybe you can use FFI to make it faster?

I tested it in MacbookPro and simulated parsing ngx.req.get_headers(), which is actually a table.

```lua
local fmt = string.format
local ffi = require("ffi")

local _M = {
    _VERSION = '1.0.0',
}
function _M:new (o)
    o = o or {}
    setmetatable(o, self)
    self.__index = self
    return o
end
function _M:add (...)
    for _, v in ipairs({ ... }) do
        table.insert(self, v)
    end
end
function _M:add_crlf ()
    self:add("\r\n")
end
function _M:add_kv_crlf (k, v)
    self:add(k, ": ", v, "\r\n")
end
function _M:len()
    local len = 0
    for _, v in ipairs(self) do
        len = len + #v
    end
    return len
end

local headers = {}
headers['accept'] = 'text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.7'
headers['sec-fetch-site'] = 'cross-site'
headers['sec-fetch-mode'] = 'navigate'
headers['host'] = 'www.google.com'
headers['sec-ch-ua'] = '"Chromium";v="122", "Not(A:Brand";v="24", "Google Chrome";v="122"'
headers['sec-ch-ua-mobile'] = '?0'
headers['user-agent'] = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/122.0.0.0 Safari/537.36'
headers['sec-ch-ua-full-version'] = '"122.0.6261.94"'
headers['accept-language'] = 'zh-CN,zh;q=0.9'
headers['sec-ch-ua-arch'] = '"x86"'
headers['content-length'] = '13'
headers['sec-ch-ua-platform'] = '"macOS"'
headers['sec-ch-ua-platform-version'] = '"14.3.1"'
headers['accept-encoding'] = 'gzip, deflate, br'
headers['sec-ch-ua-model'] = '""'
headers['referer'] = 'https://ab.com/'
headers['sec-ch-ua-bitness'] = '"64"'
headers['sec-fetch-dest'] = 'document'
headers['sec-ch-ua-wow64'] = '?0'
headers['sec-ch-ua-full-version-list'] = '"Chromium";v="122.0.6261.94", "Not(A:Brand";v="24.0.0.0", "Google Chrome";v="122.0.6261.94"'
headers['sec-fetch-user'] = '?1'
headers['upgrade-insecure-requests'] = '1'

local function t1k_concat()
    local buf = _M:new()
    buf:add(fmt("%s %s HTTP/%.1f\r\n", "GET", "/xxxx", 2.0 ))

    for k, v in pairs(headers) do
        buf:add_kv_crlf(k, v)
    end
    buf:add_crlf()
end

local function table_concat()
    local buf = {}
    buf[1] = "GET"
    buf[2] = " "
    buf[3] = "/xxxx"
    buf[4] = " "
    buf[5] = "HTTP/"
    buf[6] = tostring(2.0)
    buf[7] = "\r\n" 
    for k, v in pairs(headers) do
        buf[#buf+1] = tostring(k)
        buf[#buf+1] = ": "
        buf[#buf+1] = tostring(v) 
        buf[#buf+1] = "\r\n"
    end
    return table.concat(buf, "")
end

local function string_concat()
    local buf = "GET" .. "/xxxx" .. "HTTP/" .. tostring(2.0) .. "\r\n"
    for k, v in pairs(headers) do
        buf = buf .. tostring(k) .. ": " .. tostring(v) .. "\r\n"
    end
    buf = buf .. "\r\n"
end

local function measure_execution(func)
    collectgarbage("collect")
    local start_time = os.clock()
    local start_mem = collectgarbage("count")
    for i = 1, 1000000, 1 do
        func()
    end
    local end_time = os.clock()
    collectgarbage("collect")
    local end_mem = collectgarbage("count")
    return end_time - start_time, end_mem - start_mem
end

print("\n")
local execution_time, memory_usage = measure_execution(string_concat)
print("string concat:")
print("Execution time:", execution_time)
print("Memory usage:", memory_usage, "KB")
print("\n")

local execution_time, memory_usage = measure_execution(t1k_concat)
print("t1k concat")
print("Execution time:", execution_time)
print("Memory usage:", memory_usage, "KB")
print("\n")

local execution_time, memory_usage = measure_execution(table_concat)
print("table concat")
print("Execution time:", execution_time)
print("Memory usage:", memory_usage, "KB")
print("\n")
```

result
```
/usr/local/Cellar/openresty/1.25.3.1_1/luajit/bin/luajit test.lua

string concat:
Execution time: 1.332606
Memory usage:   4.30078125      KB


t1k concat
Execution time: 6.76907
Memory usage:   4.046875        KB


table concat
Execution time: 2.133047
Memory usage:   4.21484375      KB
```

### PR DESCRIPTION
**Pull request by C4o**
close #15

### CODE REVIEW

File: lib/resty/t1k/request.lua [hunk 1]
```diff
@@ -102,6 +102,17 @@ local function build_header()
 
     local buf = buffer:new()
     buf:add(fmt("%s %s HTTP/%.1f\r\n", ngx_req.get_method(), ngx_var.request_uri, http_version))
```
Comment (blaisewang) [inline lib/resty/t1k/request.lua]: extra line needs to be deleted

File: lib/resty/t1k/request.lua [hunk 2]
```diff
@@ -102,6 +102,17 @@ local function build_header()
...
+    buf:add(ngx_req.get_method())
+    buf:add(" ")
+    buf:add(ngx_var.request_uri)
+    buf:add(" ")
```
Comment (blaisewang) [inline lib/resty/t1k/request.lua]: HTTP/3.0 is on the way，I think `fmt(" HTTP/%.1f\r\n", http_version)` is fine

**Review by blaisewang — CHANGES_REQUESTED** [review]

File: lib/resty/t1k/request.lua (hunk 2 above)
Comment (C4o) [reply lib/resty/t1k/request.lua]: Yes, you're right.

File: lib/resty/t1k/request.lua (hunk 1 above)
Comment (C4o) [reply lib/resty/t1k/request.lua]: sorry, forgot that. I'm doing this with github editor.

**Pull request merged**